src/looprestoration.rs

```diff
@@ -136,3 +137,160 @@ decl_looprestorationfilter_fns! {
+// TODO(randomPoison): Temporarily pub until remaining looprestoration fns have
+// been deduplicated.
+#[inline(never)]
+pub(crate) unsafe fn padding<BD: BitDepth>(
+    mut dst: &mut [BD::Pixel; 70 /*(64 + 3 + 3)*/ * REST_UNIT_STRIDE],
+    mut p: *const BD::Pixel,
+    stride: ptrdiff_t,
+    mut left: *const [BD::Pixel; 4],
+    mut lpf: *const BD::Pixel,
+    mut unit_w: libc::c_int,
+    stripe_h: libc::c_int,
+    edges: LrEdgeFlags,
+) {
+    let stride = BD::pxstride(stride as usize);
+
+    let have_left = (edges & LR_HAVE_LEFT != 0) as libc::c_int;
+    let have_right = (edges & LR_HAVE_RIGHT != 0) as libc::c_int;
+
+    // Copy more pixels if we don't have to pad them
+    unit_w += 3 * have_left + 3 * have_right;
+    let mut dst_l = &mut dst[(3 * (have_left == 0) as libc::c_int) as usize..];
+    p = p.offset(-((3 * have_left) as isize));
+    lpf = lpf.offset(-((3 * have_left) as isize));
+
+    if edges & LR_HAVE_TOP != 0 {
+        // Copy previous loop filtered rows
+        let above_1 = std::slice::from_raw_parts(lpf, stride + unit_w as usize);
+        let above_2 = &above_1[stride..];
+
+        BD::pixel_copy(dst_l, above_1, unit_w as usize);
+        BD::pixel_copy(&mut dst_l[REST_UNIT_STRIDE..], above_1, unit_w as usize);
+        BD::pixel_copy(&mut dst_l[2 * REST_UNIT_STRIDE..], above_2, unit_w as usize);
+    } else {
+        // Pad with first row
+        let p = std::slice::from_raw_parts(p, unit_w as usize);
+
+        BD::pixel_copy(dst_l, p, unit_w as usize);
+        BD::pixel_copy(&mut dst_l[REST_UNIT_STRIDE..], p, unit_w as usize);
+        BD::pixel_copy(&mut dst_l[2 * REST_UNIT_STRIDE..], p, unit_w as usize);
+
+        if have_left != 0 {
+            let left = &(*left.offset(0))[1..];
+            BD::pixel_copy(dst_l, left, 3);
+            BD::pixel_copy(&mut dst_l[REST_UNIT_STRIDE..], left, 3);
+            BD::pixel_copy(&mut dst_l[2 * REST_UNIT_STRIDE..], left, 3);
+        }
+    }
+
+    let mut dst_tl = &mut dst_l[3 * REST_UNIT_STRIDE..];
+    if edges & LR_HAVE_BOTTOM != 0 {
+        // Copy next loop filtered rows
+        let below_1 = std::slice::from_raw_parts(lpf.offset(6 * stride as isize), unit_w as usize);
+        let below_2 = std::slice::from_raw_parts(lpf.offset(7 * stride as isize), unit_w as usize);
+
+        BD::pixel_copy(
+            &mut dst_tl[stripe_h as usize * REST_UNIT_STRIDE..],
+            below_1,
+            unit_w as usize,
+        );
+        BD::pixel_copy(
+            &mut dst_tl[(stripe_h + 1) as usize * REST_UNIT_STRIDE..],
+            below_2,
+            unit_w as usize,
+        );
+        BD::pixel_copy(
+            &mut dst_tl[(stripe_h + 2) as usize * REST_UNIT_STRIDE..],
+            below_2,
+            unit_w as usize,
+        );
+    } else {
+        // Pad with last row
+        let src = std::slice::from_raw_parts(
+            p.offset(((stripe_h - 1) as isize * stride as isize) as isize),
+            unit_w as usize,
+        );
```

Comment:
```suggestion
        let p = std::slice::from_raw_parts(p, (stripe_h as usize - 1) * stride + unit_w as usize);
        let src = &p[(stripe_h as usize - 1) * stride..];
```

I think it's a bit better to convert raw pointers to slices as soon as possible, and then do the rest of the operations all safely.